### PR TITLE
Update roslyn-analyzers to enable non-copyable analyzer

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Test/PdbUtilities/Shared/DummyMetadataImport.cs [hunk 1]
```diff
@@ -66,7 +66,9 @@ private void Dispose(bool disposing)
             ppvSig = (byte*)pinnedBuffer.AddrOfPinnedObject();
             pcbSig = signature.Length;
 
+#pragma warning disable RS0042 // Do not copy value
             _pinnedBuffers.Add(pinnedBuffer);
+#pragma warning restore RS0042 // Do not copy value
```
Comment (sharwell) [inline src/Test/PdbUtilities/Shared/DummyMetadataImport.cs:71]: 💭 I'm not sure if we need to exclude `GCHandle` from the analysis, but it does seem like we need to make this support customizing the set of non-copyable types from libraries.

File: eng/config/rulesets/Shipping.ruleset [hunk 2]
```diff
@@ -55,8 +56,11 @@
     <Rule Id="CA1823" Action="None" /> <!-- https://github.com/dotnet/roslyn/issues/20404: Enable rule CA1823 (remove unused field) once we have a fixer for it -->
     <Rule Id="CA1824" Action="None" /> <!-- mark assemblies with NeutralResourcesLanguageAttribute -->
     <Rule Id="CA1827" Action="None" /> <!-- https://github.com/dotnet/roslyn/issues/37959: DoNotUseCountWhenAnyCanBeUsed -->
+    <Rule Id="CA1834" Action="None" /> <!-- Consider using 'StringBuilder.Append(char)' when applicable -->
+    <Rule Id="CA1835" Action="None" /> <!-- Prefer the 'Memory'-based overloads for 'ReadAsync' and 'WriteAsync' -->
     <Rule Id="CA2000" Action="None" /> <!-- https://github.com/dotnet/roslyn/issues/25880 -->
     <Rule Id="CA2007" Action="Warning" />
+    <Rule Id="CA2100" Action="None" /> <!-- AD0001 ReviewSqlQueriesForSecurityVulnerabilities: https://github.com/dotnet/roslyn-analyzers/issues/3613 -->
```
Comment (mavasani) [inline eng/config/rulesets/Shipping.ruleset:63]: https://github.com/dotnet/roslyn-analyzers/pull/3614 should be merged and this line should be reverted. This issue is a bug in the underlying options infrastructure in the analyzers repo, which can show up in any rule which uses an analyzer config option.

File: src/Interactive/Host/Microsoft.CodeAnalysis.InteractiveHost.csproj [hunk 3]
```diff
@@ -37,6 +37,7 @@
     <Reference Include="System.Windows.Forms" />
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="..\..\Compilers\Core\Portable\InternalUtilities\NonCopyableAttribute.cs" Link="Utilities\NonCopyableAttribute.cs" />
```
Comment (mavasani) [inline src/Interactive/Host/Microsoft.CodeAnalysis.InteractiveHost.csproj:40]: I don't think injecting this attribute explicitly into every project that needs this analysis is a good approach. Can we follow the same model as PerformanceSensitiveAttribute? We should unify our approach for similar kinds of custom attribute based analyses.
https://github.com/dotnet/roslyn/blob/96af121422a90b0c63b1c675005ea9ddb4d6a2d3/eng/targets/Settings.props#L140-L146

Comment (sharwell) [reply src/Interactive/Host/Microsoft.CodeAnalysis.InteractiveHost.csproj:40]: This is what we already do for `[NonDefaultable]`. We can leave this for a follow-up for now.

Comment (mavasani) [reply src/Interactive/Host/Microsoft.CodeAnalysis.InteractiveHost.csproj:40]: I don't like building up this debt, but I'll let you and @genlu make a call here. Analyzers requiring such custom annotation attributes for analysis are only going to increase with time, so this approach of analyzer consumers having to define the attribute and include them in required project files does not look scalable.

Comment (genlu) [reply src/Interactive/Host/Microsoft.CodeAnalysis.InteractiveHost.csproj:40]: I agree with Manish, we should follow the same pattern from PerfSensitiveAnalyzer before releasing the analyzer.

Comment (sharwell) [reply src/Interactive/Host/Microsoft.CodeAnalysis.InteractiveHost.csproj:40]: This is not a public analyzer at this time. We can revisit this in the future if/when we decide to make it a public analyzer.

**Comment by sharwell** [comment]
@mavasani Updated to a new build of the analyzers and restored CA2100

File: eng/config/rulesets/Shipping.ruleset [hunk 4]
```diff
@@ -55,8 +56,11 @@
     <Rule Id="CA1823" Action="None" /> <!-- https://github.com/dotnet/roslyn/issues/20404: Enable rule CA1823 (remove unused field) once we have a fixer for it -->
     <Rule Id="CA1824" Action="None" /> <!-- mark assemblies with NeutralResourcesLanguageAttribute -->
     <Rule Id="CA1827" Action="None" /> <!-- https://github.com/dotnet/roslyn/issues/37959: DoNotUseCountWhenAnyCanBeUsed -->
+    <Rule Id="CA1834" Action="None" /> <!-- Consider using 'StringBuilder.Append(char)' when applicable -->
```
Comment (mavasani) [inline eng/config/rulesets/Shipping.ruleset:59]: Do we want to open tracking issues to enable rules which are useful, but just not in scope of this PR?

Comment (sharwell) [reply eng/config/rulesets/Shipping.ruleset:59]: I can for this one. Are there any others?

Comment (mavasani) [reply eng/config/rulesets/Shipping.ruleset:59]: CA1835, CA2249?

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for reviews on marking some value types as non-copyable

File: eng/config/rulesets/Shipping.ruleset (hunk 4 above)
Comment (sharwell) [reply eng/config/rulesets/Shipping.ruleset:59]: Issues are filed

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for reviews on marking some value types as non-copyable

**Pull request merged**